### PR TITLE
5643 refactor prescription hooks

### LINKED ISSUE #5643: Update prescriptions to use new Hook Pattern

### PR DESCRIPTION
**Pull request by CarlosNZ**
Fixes #5643

# 👩🏻‍💻 What does this PR do?

Should be no noticeable changes in user experience.

Refactor hooks using new pattern. New hooks are:

- `usePrescription` -- manages state and provides database methods for a single prescription
- `usePrescriptionList` -- provides query/data for the prescription list
- `usePrescriptionDelete` -- provides "delete" database method. Didn't incorporate into the previous hooks as deletion of prescriptions is available from both the List and Details pages
- `usePrescriptionLines` -- provides database methods for create/updates/delete of individual prescription lines. I didn't fully manage the draft state in this hook, as the logic in the existing `useDraftPrescriptionLines` is pretty complicated and interacts with a whole bunch of other things, so I left it alone and just used the hook to provide the database mutations.

## 💌 Any notes for the reviewer?
See inline comments.

# 🧪 Testing

Basically just do Prescription stuff (create, delete, edit, add/remove lines, etc) and see if anything breaks or behaves unexpectedly.

Should also test the "Prescription" component in Encounters if possible as that also uses these hooks.

### CODE REVIEW

File: client/packages/invoices/src/Prescriptions/DetailView/PrescriptionLineEdit/PrescriptionLineEditForm.tsx [hunk 1]
```diff
@@ -73,7 +73,8 @@ export const PrescriptionLineEditForm: React.FC<
   const [allocationAlerts, setAllocationAlerts] = useState<StockOutAlert[]>([]);
   const [issueQuantity, setIssueQuantity] = useState(0);
   const { format } = useFormatNumber();
-  const { items } = usePrescription.line.rows();
+  // const { items } = usePrescription.line.rows();
+  const { rows: items } = usePrescription();
```
Comment (CarlosNZ) [inline client/packages/invoices/src/Prescriptions/DetailView/PrescriptionLineEdit/PrescriptionLineEditForm.tsx:77]: There was a lot of this kind of thing, where there was a whole hook for constructing data structures that can be easily derived from the main prescription data, so I've rolled them in to the main prescription hook.

File: client/packages/invoices/src/Prescriptions/DetailView/PrescriptionLineEdit/hooks/useDraftPrescriptionLines.tsx [hunk 2]
```diff
@@ -28,13 +28,14 @@ export const useDraftPrescriptionLines = (
   item: DraftItem | null,
   date?: Date | null
 ): UseDraftPrescriptionLinesControl => {
-  const { id: invoiceId, status } = usePrescription.document.fields([
-    'id',
-    'status',
-  ]);
```
Comment (CarlosNZ) [inline client/packages/invoices/src/Prescriptions/DetailView/PrescriptionLineEdit/hooks/useDraftPrescriptionLines.tsx:34]: I don't know why there was a hook just to pull a subset of fields from an existing data structure

File: client/packages/invoices/src/Prescriptions/DetailView/PrescriptionLineEdit/hooks/useDraftPrescriptionLines.tsx [hunk 3]
```diff
@@ -28,13 +28,14 @@ export const useDraftPrescriptionLines = (
...
+  const {
+    query: { data: prescriptionData },
+  } = usePrescription();
+  const { id: invoiceId, status } = prescriptionData ?? {};
 
-  const { data: lines, isLoading: prescriptionLinesLoading } =
-    usePrescription.line.stockLines(item?.id ?? '');
+  const lines = prescriptionData?.lines.nodes.filter(
+    line => item?.id === line.item.id
+  );
```
Comment (CarlosNZ) [inline client/packages/invoices/src/Prescriptions/DetailView/PrescriptionLineEdit/hooks/useDraftPrescriptionLines.tsx:38]: Again, derivable from main prescription data

File: client/packages/invoices/src/Prescriptions/DetailView/Toolbar.tsx [hunk 4]
```diff
@@ -12,28 +12,58 @@ import {
...
+    canDelete: !isDisabled,
+    messages: {
+      confirmMessage: t('messages.confirm-delete-lines', {
+        count: selectedRows.length,
+      }),
+      deleteSuccess: t('messages.deleted-lines', {
+        count: selectedRows.length,
+      }),
+    },
+  });
```
Comment (CarlosNZ) [inline client/packages/invoices/src/Prescriptions/DetailView/Toolbar.tsx:65]: This was in a hook before, but I think it makes more sense to have UI stuff in the UI components and only have logic and database interactions in hooks.

File: client/packages/invoices/src/Prescriptions/api/hooks/usePatchState.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
+/**
+ * Simple hook to manage a data patch state, which is then passed into a
+ * mutation to update a record.
+ *
+ * It holds the patch state, as well as provides methods to update and reset the
+ * patch, and provides the "dirty" value (i.e. whether or not there is any
+ * difference between the patch and the original "reference" data)
+ */
+
```
Comment (CarlosNZ) [inline client/packages/invoices/src/Prescriptions/api/hooks/usePatchState.ts:9]: This is exactly the same in several places, so made a new hook for this. I'll make a follow-up PR (this is already big enough!) where I've moved this into "common" and implemented in other places that follow the same pattern.

File: client/packages/invoices/src/Prescriptions/api/hooks/usePrescriptionLines.ts [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    draftPrescriptionLines: DraftStockOutLine[];
+    patch?: RecordPatch<PrescriptionRowFragment>;
+  }) => {
+    if (patch && id !== '') patch.id = id;
+    const input = {
+      insertPrescriptionLines: draftPrescriptionLines
+        .filter(
+          ({ type, isCreated, numberOfPacks }) =>
+            isCreated &&
+            type === InvoiceLineNodeType.StockOut &&
```
Comment (jmbrunskill) [inline client/packages/invoices/src/Prescriptions/api/hooks/usePrescriptionLines.ts:82]: Any reason why they wouldn't be the right type here?

**Review by jmbrunskill — APPROVED** [review]
Look great Carl!
Appreciate all the comments and explainations.

I haven't tested out the jsonforms stuff but assume since you've just been working on that area you'd have a better idea of how that works.

File: client/packages/invoices/src/Prescriptions/DetailView/DetailView.tsx [hunk 7]
```diff
@@ -44,7 +45,7 @@ export const PrescriptionDetailView: FC = () => {
     setMode(ModalMode.Create);
   };
 
-  if (isLoading) return <DetailViewSkeleton hasGroupBy={true} hasHold={true} />;
+  if (loading) return <DetailViewSkeleton hasGroupBy={true} hasHold={true} />;
```
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply client/packages/invoices/src/Prescriptions/DetailView/DetailView.tsx:48]: No, it's inconsistent. I think we've been using `isLoading` but Tanstack returns `loading`, so sometimes we've passed it through directly, other times renamed 🤷‍♂️

File: client/packages/invoices/src/Prescriptions/api/hooks/document/usePrescriptionFields.ts [hunk 8]
```diff
@@ -1,49 +0,0 @@
...
-  const updateAndCaptureError: typeof api.update = async input => {
-    const result = await api.update(input);
-    const errors =
-      result?.updatePrescriptions?.flatMap(({ response }) => {
-        if ('error' in response) return [response.error];
-        return [];
-      }) || [];
-
-    for (const error of errors) {
-      if (error.__typename == 'InvalidStockSelection') {
```
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply client/packages/invoices/src/Prescriptions/api/hooks/document/usePrescriptionFields.ts:33]: Yeah, this was a special case -- most of the mutations didn't return specific errors, just a generic throw, which is what they all now use. I think we should keep an eye on this and go back and add more specific error notifications if we feel they're required. But for the most part, there hasn't been any change regarding this in this PR.

File: client/packages/invoices/src/Prescriptions/api/hooks/usePrescriptionLines.ts (hunk 6 above)
Comment (CarlosNZ) [reply client/packages/invoices/src/Prescriptions/api/hooks/usePrescriptionLines.ts:82]: Not that I can think of, but this is just copied over from the original. Worth removing you think?

File: client/packages/invoices/src/Prescriptions/DetailView/PrescriptionLineEdit/hooks/useDraftPrescriptionLines.tsx (hunk 3 above)
Comment (CarlosNZ) [reply client/packages/invoices/src/Prescriptions/DetailView/PrescriptionLineEdit/hooks/useDraftPrescriptionLines.tsx:38]: Yeah, I think the invalidation cases are all covered now, and easier to keep track of. Let's keep an eye out ;)
